File: day13.cpp
#include<iostream>
#include<fstream>
#include<iomanip>
#include<vector>
#include"functions.h"

std::string make_list(std::string input){
    input.insert(input.begin(),'[');

    std::string::iterator itr = input.begin();

    do{itr++;}while(*itr != ',' and *itr != ']');

    input.insert(itr,']');

    return input;
}

std::string extract_element(std::string& input){
    int level = 0;
    std::string::iterator itr = input.begin();

    while(itr < input.end()){
        if(*itr == '['){level++;}else if(*itr == ']'){level--;}
        itr++;
        if(*itr == ',' and level == 0){
            std::string tmp = std::string(input.begin(),itr);
            input.erase(input.begin(),itr);
            return tmp;
        }
    }
    std::string tmp = input;
    input = "";
    return tmp;
}

void compare(std::string& left, std::string& right, bool& result){
    // We compare the packets left and right recursively; one call to compare() compares the outermost bracket
    // in left and right

    // removing the outermost brackets
    left.erase(left.begin()); right.erase(right.begin());

    while(left.size() > 0 and right.size() > 0){
        // both packets contain elements to compare. We deal with the two next elements and remove them afterwards.
        // the possibilities are: Two lists, one list in either left or right or to integers

        if(left[0] == '[' and right[0] == '['){
            // the next elements are two lists, so we go one recursion deeper
            left.erase(left.begin()); right.erase(right.begin());
            compare(left,right,result);
        }
    }

    if(left.size() > 0){
        // the left packet contains more elements than the right one, the packets are not ordered
        result = false;
    }
}

int main(){
    std::fstream file("day13_packets_test.txt"); if(not file.is_open()){std::cout << "File not open!" << std::endl; return -1;}

    std::string left, right, tmp;
    int pair_index = 1,sum = 0;
    bool comparison;
    while(std::getline(file,left) and std::getline(file,right)){
        // the next two lines are contained in the strings left and right
        left.erase(left.begin()); right.erase(right.begin());
        left.erase(left.end()-1); right.erase(right.end()-1);
        
        std::cout << "pair " << std::setw(3) << pair_index << "" << std::endl;
        // let's compare and, if ordered, add the corresponding pair index to the sum:
        comparison = true;
        // compare(left,right,comparison);
        if(comparison){
            sum += pair_index;

            // std::cout << left << std::endl << right << right << std::endl << std::endl;
        }


        // fetching the empty line
        std::getline(file,tmp);
        pair_index++;
    }

    file.close();

    std::cout << "Sum of components = " << sum << std::endl;

    return 0;
}

File: day13.py
import numpy as np

def packetprep(packetstring):
    """
    Takes a `string` and returns a list containing all values as well as the brackets.
    """
    packetstring = packetstring.replace("[",",[,")
    packetstring = packetstring.replace("]",",],")
    packetstring = packetstring.replace(",,",",")
    packetstring = packetstring.replace(",,",",")
    packet = packetstring.split(",")[1:-1]

    for i in range(len(packet)):
        if packet[i] not in ("[","]"):
            packet[i] = int(packet[i])

    return packet

def packetcomp(iL,iR,left_packet,right_packet):
    """
    Recursively compare two packets. `iL` and `iR` mark the indices we are currently comparing. We check all scenarios consecutively:
    * Are there `[`s at both indices, i.e. did we find two lists? Then increment `iL` and `iR` and continue.
    * Is there a `[` at only one of the two indices? Then turn the value at the other index into a list (by
    inserting `[` and `]` before and after, respectively) and continue.
    * Is one list ending before the other? Then return `True` or `False` depending on which list ended first.
    * Are both lists ending? Then increment `iL` and `iR` and continue.
    
    If all of the above doesn't apply, we're simply comparing two numbers.

    Note that currently, the behaviour is undefined when no decision can be made - this case was not mentioned
    in the problem description.
    """

    if left_packet[iL] == "[" and right_packet[iR] == "[":
        # a list in both packets
        return packetcomp(iL+1,iR+1,left_packet,right_packet)
    
    if left_packet[iL] != "]" and right_packet[iR] == "]":
        # the right list ended before the left list - they are not in order
        return False
    
    if left_packet[iL] == "]" and right_packet[iR] != "]":
        # the left list ended before the left list - they are in order
        return True
    
    if left_packet[iL] == "[" and right_packet[iR] != "[":
        # a list in the left packet - we need to turn the element in the right packet into a list
        right_packet = right_packet[:iR] + ["[",] + [right_packet[iR],] + ["]",] + right_packet[iR+1:]
        return packetcomp(iL,iR,left_packet,right_packet)
    
    if left_packet[iL] != "[" and right_packet[iR] == "[":
        # a list in the right packet - we need to turn the element in the left packet into a list
        left_packet = left_packet[:iL] + ["[",] + [left_packet[iL],] + ["]",] + left_packet[iL+1:]
        return packetcomp(iL,iR,left_packet,right_packet)
    
    if left_packet[iL] == "]" and right_packet[iR] == "]":
        # a list in both packets ends
        return packetcomp(iL+1,iR+1,left_packet,right_packet)

    # if we got to this point, both entries are numbers which we can compare normally

    if left_packet[iL] == right_packet[iR]:
        # the numbers are equal, we cannot make a decision here
        return packetcomp(iL+1,iR+1,left_packet,right_packet)
    
    # the numbers are uneqal - now we're just comparing
    if left_packet[iL] < right_packet[iR]:
        return True
    else:
        return False

# opening the file
file = open("day13_packets.txt","r")

line = "dummy"
pairindex = 0
result = 0

while(line != ""):
    # reading in the packet strings
    left_packet  = packetprep(file.readline())
    right_packet = packetprep(file.readline())
    line = file.readline()
    pairindex += 1

    # comparing the packets:
    if packetcomp(0,0,left_packet,right_packet):
        # print("Pair {}: Right order".format(pairindex))
        result += pairindex
    # else:
    #     print("Pair {}: Failed".format(pairindex))

print("The result is {}".format(result))

file.close()